protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java

```diff
@@ -107,23 +110,38 @@ public T parse(InputStream stream) {
             }
           }
         }
+        CodedInputStream cis;
+        if (stream instanceof ReadableBuffer) {
+          // This is only a temporary solution, and deserves a better solution.  We copy here to
+          // avoid Protobuf's bad deserialization buffering.  In reality, we should be able to pass
+          // a collection of ByteBuffers to the decoder and avoid buffering at all.  The code below
+          // should eventually be removed in favor of a proper solution.
+          ReadableBuffer buffer = (ReadableBuffer) stream;
+          // copyBuffer must never leave scope and may never be modified.
+          ByteBuffer copyBuffer = ByteBuffer.allocate(buffer.readableBytes());
+          buffer.readBytes(copyBuffer);
+          // Unsafe is used here because we are very careful about how we use it.  It must not be
+          // used without careful scrutiny.
+          cis = UnsafeByteOperations.unsafeWrap(copyBuffer).newCodedInput();
```

Comment:
Guys I don't get what's so difficult?  Look you can perform it as follows, were the last element of the array is a CRC32 of the previous array elements:

```Java
import java.util.zip.CRC32;

public class ImmutableArray {

  public static void main(String[] args) {

    CRC32 crc = new CRC32();
    final byte [] array = {1,2,3,0};
    	
    crc.reset();
    crc.update(array, 0, array.length-2); 

    array[array.length - 1] = (byte) crc.getValue();
    System.out.println();
    System.out.println("Update the first element with the same previous value");
    array[0] = 1;
    	  
    crc.reset();
    crc.update(array, 0, array.length-2);

   	System.out.println( "Has the array changed? " + (((byte) array[3])  !=  ((byte) crc.getValue()))  );
   	System.out.println();

   	System.out.println("Update the first element with a different value");
   	array[0] = 10;
    	  
    crc.reset();
    crc.update(array, 0, array.length-2);

   	System.out.println( "Has the array changed? " + (((byte) array[3])  !=  ((byte) crc.getValue()))  );

  }

}
```

Below is the output:

```
> javac ImmutableArray.java && java ImmutableArray

Update the first element with the same previous value
Has the array changed? false

Update the first element with a different value
Has the array changed? true

>
```